### PR TITLE
Change Governor.cancel to receive all parameters

### LINKED ISSUE #4052: Include missing _cancel(uint256 proposal_id) in GovernorTimelockControl
**🧐 Motivation**
When using the Governor contract with the GovernorTimelockControl, one needs to override ```cancel(uint256 proposalId)``` because both Pending and Queued proposals must be cancellable. The problem is that it's not possible to invoke the inherited function ```_cancel(address[] memory targets, uint256[] memory values, bytes[] memory calldatas, bytes32 descriptionHash``` from inside the overidden public cancel function.
    )

**📝 Details**
A solution would be to let the GovernorTimelockControl override both the internal _cancel functions it inherits from Governor, in which case the four-argument function simply forwards to the single-argument function (as is the case in the Governor contract). The logic currently inside _cancel would move to the new single-argument function.

The downside is that it's a breaking change for anyone using the GovernorTimelockControl.
I will submit a PR with a link to this issue.

### PR DESCRIPTION
**Pull request by frangio**
Fixes #4052 

This PR implements option 1 from https://github.com/OpenZeppelin/openzeppelin-contracts/issues/4052#issuecomment-1435354280.

Still open to other options but this seems the best one by far.

#### PR Checklist

- [x] Tests
- [x] Documentation
- [ ] Changeset entry (run `npx changeset add`)
  - No changeset because `cancel` is still unreleased.

### CODE REVIEW

File: test/helpers/governance.js [hunk 1]
```diff
@@ -63,18 +63,15 @@ class GovernorHelper {
...
-    switch (visibility) {
-      case 'external':
-        return this.governor.methods['cancel(uint256)'](...concatOpts([proposal.id], opts));
-      case 'internal':
-        return this.governor.methods['$_cancel(address[],uint256[],bytes[],bytes32)'](
-          ...concatOpts(proposal.shortProposal, opts),
-        );
-      default:
-        throw new Error(`unsuported visibility "${visibility}"`);
+    if (proposal.useCompatibilityInterface) {
```
[the remark this answers is not in the document]
Comment (Amxx) [reply test/helpers/governance.js:69]: I would have done:

```js
  cancel(visibility = 'external', opts = null) {
    const proposal = this.currentProposal;

    switch (visibility) {
      case 'external':
        if (proposal.useCompatibilityInterface) {
          return this.governor.methods['cancel(uint256)'](...concatOpts([proposal.id], opts));
        } else {
          return this.governor.methods['cancel(address[],uint256[],bytes[],bytes32)'](
            ...concatOpts(proposal.shortProposal, opts),
          );
        }
      case 'internal':
        return this.governor.methods['$_cancel(address[],uint256[],bytes[],bytes32)'](
          ...concatOpts(proposal.shortProposal, opts),
        );
      default:
        throw new Error(`unsuported visibility "${visibility}"`);
    }
  }
```

**Review by Amxx — COMMENTED** [review]
I have some doubts about the testing part (see comments).

File: contracts/governance/Governor.sol [hunk 2]
```diff
@@ -336,10 +336,16 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor, IERC721Receive
     /**
      * @dev See {IGovernor-cancel}.
      */
-    function cancel(uint256 proposalId) public virtual override {
+    function cancel(
```
[the remark this answers is not in the document]
Comment (frangio) [reply contracts/governance/Governor.sol:339]: Yes definitely wasteful but I didn't find a way around it.

File: test/helpers/governance.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (frangio) [reply test/helpers/governance.js:69]: That's what I did initially.

> How do we make sure it works as expected in all the cases where the public ones revert ?

But we don't have tests for these cases.

File: contracts/governance/compatibility/GovernorCompatibilityBravo.sol [hunk 3]
```diff
@@ -77,37 +77,63 @@ abstract contract GovernorCompatibilityBravo is IGovernorTimelock, IGovernorComp
...
+        ) = _getProposalParameters(proposalId);
+
+        cancel(targets, values, calldatas, descriptionHash);
+    }
+
+    /**
+     * @dev Cancel a proposal with GovernorBravo logic. At any moment a proposal can be cancelled, either by the
+     * proposer, or by third parties if the proposer's voting power has dropped below the proposal threshold.
+     */
+    function cancel(
```
[the remark this answers is not in the document]
Comment (frangio) [reply contracts/governance/compatibility/GovernorCompatibilityBravo.sol:122]: It potentially violates checks-effects-interactions... `_cancel` is making external calls in some cases, they are to trusted contracts, but I don't feel comfortable with this.

File: test/helpers/governance.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Amxx) [reply test/helpers/governance.js:69]: We do have quite extensive tests for both the internal and the public versions:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/test/governance/Governor.test.js#L397-L502

[the remark this answers is not in the document]
Comment (frangio) [reply test/helpers/governance.js:69]: I see that those are the tests that are failing here. Didn't realize that when I made this change.

**Comment by frangio** [comment]
@Amxx Tests are using the internal function now. Please merge if it looks ok.

**Review by Amxx — APPROVED** [review]
LGTM

**Pull request merged**